File: web/src/mediaKind.js
/** Clasificación de adjuntos de chat (web) — alineada con backend uploads.js */

const VIDEO_EXT = /\.(mp4|mov|webm|mkv|avi|m4v|3gp)$/i;
const IMAGE_EXT = /\.(jpe?g|png|gif|webp|jfif|bmp)$/i;
const AUDIO_EXT = /\.(webm|ogg|mp3|m4a|wav|aac|opus)$/i;

export function extOf(name = '') {
  const m = String(name).toLowerCase().match(/\.[a-z0-9]+$/);
  return m ? m[0] : '';
}

export function isVideoMessage(message) {
  if (!message) return false;
  if (message.type === 'video') return true;
  const mime = (message.mediaMime || '').toLowerCase();
  if (mime.startsWith('video/')) return true;
  return VIDEO_EXT.test(message.mediaName || '');
}

export function isImageMessage(message) {
  if (!message) return false;
  if (message.type === 'image') return true;
  const mime = (message.mediaMime || '').toLowerCase();
  if (mime.startsWith('image/')) return true;
  return IMAGE_EXT.test(message.mediaName || '');
}

export function isAudioMessage(message) {
  if (!message) return false;
  if (message.type === 'audio') return true;
  const mime = (message.mediaMime || '').toLowerCase();
  if (mime.startsWith('audio/')) return true;
  return AUDIO_EXT.test(message.mediaName || '');
}

/** type para FormData al subir */
export function classifyUploadFile(file) {
  if (!file) return 'file';
  const mime = (file.type || '').toLowerCase();
  const name = file.name || '';
  if (mime.startsWith('image/') || IMAGE_EXT.test(name)) return 'image';
  if (mime.startsWith('audio/') || AUDIO_EXT.test(name)) return 'audio';
  if (mime.startsWith('video/') || VIDEO_EXT.test(name)) return 'video';
  return 'file';
}

export function formatBytes(n) {
  const bytes = Number(n) || 0;
  if (bytes < 1024) return `${bytes} B`;
  if (bytes < 1024 * 1024) return `${Math.round(bytes / 1024)} KB`;
  return `${(bytes / (1024 * 1024)).toFixed(1)} MB`;
}

export function fileKindIcon(name = '', mime = '') {
  const n = String(name).toLowerCase();
  const m = String(mime).toLowerCase();
  if (m.startsWith('video/') || VIDEO_EXT.test(n)) return '🎬';
  if (m.includes('pdf') || n.endsWith('.pdf')) return '📄';
  if (/\.(doc|docx)$/.test(n) || m.includes('word')) return '📝';
  if (/\.(xls|xlsx|csv)$/.test(n) || m.includes('sheet') || m.includes('excel')) return '📊';
  if (/\.(ppt|pptx)$/.test(n) || m.includes('presentation')) return '📑';
  if (/\.(zip|rar|7z|gz|tar)$/.test(n) || m.includes('zip') || m.includes('rar') || m.includes('7z')) {
    return '🗜️';
  }
  if (m.startsWith('audio/') || AUDIO_EXT.test(n)) return '🎵';
  return '📎';
}

/** accept= para input de documentos / archivos generales */
export const DOC_ACCEPT =
  '.pdf,.doc,.docx,.xls,.xlsx,.ppt,.pptx,.txt,.csv,.zip,.rar,.7z,.gz,.tar,.mp4,.mov,.webm,.mkv,.avi,.m4v,image/*,audio/*,video/*,application/pdf,application/zip,application/x-rar-compressed,application/vnd.rar,application/x-7z-compressed';

export const VIDEO_ACCEPT = 'video/*,.mp4,.mov,.webm,.mkv,.avi,.m4v,.3gp';

File: mobile/lib/media_kind.dart
// Clasificación de adjuntos (app) — alineada con backend uploads.js

bool isImageMedia({String? type, String? mime, String? name}) {
  if (type == 'image') return true;
  final m = (mime ?? '').toLowerCase();
  if (m.startsWith('image/')) return true;
  return RegExp(r'\.(jpe?g|png|gif|webp|jfif|bmp)$', caseSensitive: false)
      .hasMatch(name ?? '');
}

bool isVideoMedia({String? type, String? mime, String? name}) {
  if (type == 'video') return true;
  final m = (mime ?? '').toLowerCase();
  if (m.startsWith('video/')) return true;
  return RegExp(r'\.(mp4|mov|webm|mkv|avi|m4v|3gp)$', caseSensitive: false)
      .hasMatch(name ?? '');
}

bool isAudioMedia({String? type, String? mime, String? name}) {
  if (type == 'audio') return true;
  final m = (mime ?? '').toLowerCase();
  if (m.startsWith('audio/')) return true;
  return RegExp(r'\.(webm|ogg|mp3|m4a|wav|aac|opus)$', caseSensitive: false)
      .hasMatch(name ?? '');
}

/// Valor de `type` para multipart (video se envía como video; backend guarda file).
String classifyUploadName(String name, {String? mime}) {
  if (isImageMedia(mime: mime, name: name)) return 'image';
  if (isAudioMedia(mime: mime, name: name)) return 'audio';
  if (isVideoMedia(mime: mime, name: name)) return 'video';
  return 'file';
}

String fileKindEmoji(String? name, String? mime) {
  final n = (name ?? '').toLowerCase();
  final m = (mime ?? '').toLowerCase();
  if (isVideoMedia(mime: mime, name: name)) return '🎬';
  if (n.endsWith('.apk') || m.contains('android.package')) return '📦'; // histórico OTA/recibidos; no se envían por chat
  if (m.contains('pdf') || n.endsWith('.pdf')) return '📄';
  if (n.endsWith('.doc') || n.endsWith('.docx') || m.contains('word')) return '📝';
  if (n.endsWith('.xls') || n.endsWith('.xlsx') || n.endsWith('.csv')) return '📊';
  if (RegExp(r'\.(zip|rar|7z|gz|tar)$').hasMatch(n)) return '🗜️';
  return '📎';
}

/// Extensiones permitidas en «Documento / archivo» (estilo WhatsApp).
/// Sin .apk: distribución solo por OTA (`app-updates`).
final _docExtRe = RegExp(
  r'\.(pdf|docx?|xlsx?|xlsm|pptx?|ppsx?|txt|text|rtf|odt|ods|odp|csv|md|log|json|xml|html?|zip|rar|7z|gz|tar|epub|azw3?)$',
  caseSensitive: false,
);

const kDocumentExtensions = <String>[
  'pdf',
  'doc',
  'docx',
  'xls',
  'xlsx',
  'xlsm',
  'ppt',
  'pptx',
  'pps',
  'ppsx',
  'txt',
  'rtf',
  'odt',
  'ods',
  'odp',
  'csv',
  'md',
  'log',
  'json',
  'xml',
  'html',
  'htm',
  'zip',
  'rar',
  '7z',
  'gz',
  'tar',
  'epub',
];

bool isDocumentFile({String? mime, String? name}) {
  final n = name ?? '';
  final m = (mime ?? '').toLowerCase();
  if (n.toLowerCase().endsWith('.apk') || m.contains('android.package')) {
    return false;
  }
  if (_docExtRe.hasMatch(n)) return true;
  if (m.startsWith('text/')) return true;
  if (m.contains('zip') || m.contains('x-rar') || m.contains('x-7z')) return true;
  if (m.contains('pdf') ||
      m.contains('msword') ||
      m.contains('wordprocessingml') ||
      m.contains('spreadsheetml') ||
      m.contains('presentationml') ||
      m.contains('ms-excel') ||
      m.contains('ms-powerpoint') ||
      m.contains('opendocument') ||
      m.contains('rtf')) {
    return true;
  }
  return false;
}

String formatBytes(int? n) {
  final bytes = n ?? 0;
  if (bytes < 1024) return '$bytes B';
  if (bytes < 1024 * 1024) return '${(bytes / 1024).round()} KB';
  return '${(bytes / (1024 * 1024)).toStringAsFixed(1)} MB';
}
